src/Common/DNSResolver.cpp

```diff
@@ -138,16 +140,27 @@ static DNSResolver::IPAddresses resolveIPAddressImpl(const std::string & host)
     return addresses;
 }
 
-static String reverseResolveImpl(const Poco::Net::IPAddress & address)
+static std::vector<String> reverseResolveImpl(const Poco::Net::IPAddress & address)
 {
-    Poco::Net::SocketAddress sock_addr(address, 0);
-
-    /// Resolve by hand, because Poco::Net::DNS::hostByAddress(...) does getaddrinfo(...) after getnameinfo(...)
-    char host[1024];
-    int err = getnameinfo(sock_addr.addr(), sock_addr.length(), host, sizeof(host), nullptr, 0, NI_NAMEREQD);
-    if (err)
-        throw Exception("Cannot getnameinfo(" + address.toString() + "): " + gai_strerror(err), ErrorCodes::DNS_ERROR);
-    return host;
+    boost::asio::io_service io_service;
+    YukiWorkshop::DNSResolver resolver(io_service);
+
+    std::vector<std::string> ptr_records;
+
+    resolver.resolve_a4ptr(boost::asio::ip::address_v4::from_string(address.toString()), [&](int err, auto & hosts, auto &, auto &, uint) {
+       if (err) {
+           throw Exception("Cannot resolve: " + address.toString() + YukiWorkshop::DNSResolver::error_string(err), ErrorCodes::DNS_ERROR);
+       }
+
+       for (auto & it : hosts) {
+           ptr_records.emplace_back(it);
+       }
+
+    });
+
+    io_service.run();
+
+    return ptr_records;
 }
```

Comment:
We need a library that returns multiple PTR records for a given ip address. The problem is that most libraries return only one, including `getnameinfo`. I am also not a big fan of the namespace, but that's the only working library I could find.

The easiest path is to keep `cpp-dns`, but I believe we could get rid of it if we implemented something similar, it'll just take more time.

Reply:
I wouldn't consider it to be useless. It's much more idiomatic and easier to use than libudns. libudns is cryptic and hard to use, I believe that we need a wrapper. The question is if we want to build a wrapper or use an existing one.

Reply:
If you want, I can implement the wrapper myself and get rid of cpp-dns. Would that be enough?

Reply:
The only "idiomatic" thing it does is calling `dns_new`/`dns_init`/`dns_open` in ctor and `dns_free` in dtor of `DNSResolver` class. And it does this thing wrong and not exception-safe way... Btw, we have our own class named `DNSResolver` in `DB` namespace. I suppose it would be more optimal to call these functions in global instance of `DB::DNSResolver` instead of constructing temporary resolver each time.

Moreover, this library unconditionally prints debug output into stderr. It means that we cannot use this library as is.

Also I don't understand why do we need `boost::asio` here (was it the topic of a workshop for students or something?). These usages of `io_service` and lambdas in `reverseResolveImpl` for no reason look cryptic too.

So, this wrapper works incorrectly and tries to do stuff that we don't need. Meanwhile there are only about 500 lines of code (most of that we don't need). Yes, let's build our own 50-lines wrapper.

Reply:
Hi. Just to confirm, you want those init functions to be called during the initialization of DB::DNSResolver, right? I guess de-initialization would never happen/ be needed in that case. I wonder if we'd have any issues with threads in that sense.

Reply:
> Just to confirm, you want those init functions to be called during the initialization of DB::DNSResolver, right?

I want it to be implemented an optimal way. It's up to you to decide what way is optimal enough and why.

> I guess de-initialization would never happen/ be needed in that case.

In that case deinitialization must happen on server startup. Maybe it also makes sense to reinitialize it on DNS cache update/drop, but I'm not sure, need to check udns docs.

> I wonder if we'd have any issues with threads in that sense.

Waht issues, for example? And with what threads? Does udns run some threads?

Reply:
I just tested `gethostbyaddr_r` with the same setup and I get a single PTR record in `h_name`, while `h_alias` is null. I believe alias refers to CNAME records, which are not the same as PTR records. Not sure about it, tho. Plus, according to the docs, `gethostbyaddr` is deprecated and `getnameinfo` should be used instead. `getnameinfo` is what ClickHouse uses nowadays and it is limited to one PTR record.

Reply:
Ok, seems like `gethostbyaddr_r` behavior depends on environment. It returns all PTR records on Manjaro with glibc 2.35, but now I tested on Ubuntu 20.04 with glibc 2.31 and it returned only one record. So we cannot rely on `gethostbyaddr_r` and adding extra library for this is justified. Now we need to decide which one is better: `udns` or `c-ares` (or maybe some other library). According to [the changelog](https://github.com/c-ares/c-ares/blob/main/CHANGES.0) `c-ares` supports multiple PTR records since 2008. And `c-ares` looks much more popular than `udns`.

Reply:
That's interesting, I have just tested c-ares `ares_getnameinfo` and got only one PTR record. Maybe it's a different method?

Reply:
Where exactly did you see that? As far as my digging and experimentation went, I could only find the call I mentioned

Reply:
Sorry, I mean `ares_gethostbyaddr`, not `ares_getnameinfo`

Reply:
Ok, I just tested `ares_gethostbyaddr` and it indeed works. I will build the wrapper around it then, ok?

Reply:
ClickHouse already builds c-ares because grpc depends on it. I believe that whenever -DENABLE_GRPC is set to ON, c-ares is built as well. Adding c-ares as a submodule as it is fails because it tries to add yet another target called c-ares and it conflicts with the one built by grpc. Based on that, I see a couple of paths forward:

1. Build c-ares as a submodule and link grpc against it. Requires changes in grpc, which is not in the scope of this PR, but it would be necessary.
2. Link clickhouse_common_io against c-ares built by grpc. If c-ares target doesn't exist because grpc wasn't built, build c-ares submodule. Dependecy graph becomes a bit messy and not clear, but leaves grpc untouched.
3. Build c-ares as a submodule, but with different name. I don't see this one as a good option, but leaving it on the table.

Which one do you guys prefer and what do you suggest?

Btw, I have a draft implementation of a c-ares wrapper and already tested it on clickhouse with approach #2 (without the conditional build, tho). It worked!

Reply:
Cool, so we don't need to add extra libraries! The first option is the best, it makes sense to try it first.